hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/ConsistentBucketIdentifier.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.index.bucket;
+
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.ConsistentHashingNode;
+import org.apache.hudi.common.model.HoodieConsistentHashingMetadata;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.util.hash.HashID;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+public class ConsistentBucketIdentifier extends BucketIdentifier {
+
+  /**
+   * Hashing metadata of a partition
+   */
+  private final HoodieConsistentHashingMetadata metadata;
+  /**
+   * In-memory structure to speed up ring mapping (hashing value -> hashing node)
+   */
+  private final TreeMap<Integer, ConsistentHashingNode> ring;
+  /**
+   * Mapping from fileId -> hashing node
+   */
+  private final Map<String, ConsistentHashingNode> fileIdToBucket;
+
+  public ConsistentBucketIdentifier(HoodieConsistentHashingMetadata metadata) {
+    this.metadata = metadata;
+    this.fileIdToBucket = new HashMap<>();
+    this.ring = new TreeMap<>();
+    initialize();
+  }
+
+  public Collection<ConsistentHashingNode> getNodes() {
+    return ring.values();
+  }
+
+  public HoodieConsistentHashingMetadata getMetadata() {
+    return metadata;
+  }
+
+  public int getNumBuckets() {
+    return getNodes().size();
+  }
+
+  /**
+   * Get bucket of the given file group
+   *
+   * @param fileId the file group id. NOTE: not filePfx (i.e., uuid)
+   * @return
+   */
+  public ConsistentHashingNode getBucketByFileId(String fileId) {
+    return fileIdToBucket.get(fileId);
+  }
+
+  public ConsistentHashingNode getBucket(HoodieKey hoodieKey, String indexKeyFields) {
+    return getBucket(getHashKeys(hoodieKey, indexKeyFields));
+  }
+
+  protected ConsistentHashingNode getBucket(List<String> hashKeys) {
+    int hashValue = 0;
+    for (int i = 0; i < hashKeys.size(); ++i) {
+      hashValue = HashID.getXXHash32(hashKeys.get(i), hashValue);
```

Comment:
I followed the standard List<?>.hashCode() implementation, which make sequential hash function call to each element. Concatenating keys also introduce additional costs (meanly memory).

Reply:
I see your point. My concern is primarily that this chaining is unnecessary for our use-case and plants perf trap for ourselves in the future: this method lies on the hot-path (imagine bulk-inerting 1B rows) and therefore we want to make sure every operation is as lean as possible w/ minimal overhead.

I don't see any structural reason (like enhanced collision resistance for ex) required in this case, so i'd suggest to avoid chained hashing and instead do a hash from concatenated string. String concatenation does have its overhead but it's considerably less than chained hashing invocation

Reply:
I did a local micro benchmark and found the standard hashCode is best regarding performance... Not sure how they performs with respect to other aspects like collision.  

Since the hashing affects the data distribution and it will not be easily changed after we publish it, we should be more careful. Do you have any further advice on it? For now, I would choose concatenation way if we keep using the XXHash32 algo.

The benchmark result with 10,000,000 calls to these hash functions:
- Test1: single key="abc";
  - obj.hashCode(): 24ms, chained hash: 1914 ms, concatenate hash: 2282ms
- Test2: several small keys=("abc", "def", "ghi", "jkl", "mno", "pqr")
  - obj.hashCode(): 155ms, chained hash: 8024ms, concatenate hash: 3252ms
- Test3: two long keys=(UUID, UUID)
  - obj.hashCode(): 59ms, chained hash: 3564ms, concatenate hash: 4170ms
- Test4: several long keys=(UUID, UUID, UUID, UUID, UUID)
  - obj.hashCode(): 213ms, chained hash: 8157ms, concatenate hash: 6179ms